pinot-core/src/main/java/com/linkedin/pinot/core/startree/OffHeapStarTreeBuilder.java

```diff
@@ -15,893 +15,870 @@
+    for (String dimensionName : _dimensionNames) {
+      if (!dimensionName.equals(_timeColumnName)
+          && _dictionaryMap.get(dimensionName).size() <= _skipMaterializationCardinalityThreshold) {
         defaultSplitOrder.add(dimensionName);
       }
     }
-    if (timeColumnName != null) {
-      defaultSplitOrder.remove(timeColumnName);
-    }
+    // Sort in descending order.
     Collections.sort(defaultSplitOrder, new Comparator<String>() {
       @Override
       public int compare(String o1, String o2) {
-        return dictionaryMap.get(o2).size() - dictionaryMap.get(o1).size(); // descending
+        return _dictionaryMap.get(o2).size() - _dictionaryMap.get(o1).size();
       }
     });
     return defaultSplitOrder;
   }
 
-  private Set<String> computeDefaultDimensionsToSkipMaterialization() {
-    Set<String> skipDimensions = new HashSet<String>();
-    for (String dimensionName : dimensionNames) {
-      if (dictionaryMap.get(dimensionName).size() > skipMaterializationCardinalityThreshold) {
-        skipDimensions.add(dimensionName);
-      }
-    }
-    return skipDimensions;
-  }
-
-  /*
-   * Sorts the file on all dimensions
+  /**
+   * Helper method to sort the documents using the mmap buffer and according to the sort order.
+   *
+   * @param startDocId start document id.
+   * @param endDocId end document id.
+   * @throws Exception
    */
-  private void sort(File file, int startDocId, int endDocId) throws IOException {
-    if (debugMode) {
-      LOG.info("BEFORE SORTING");
-      printFile(file, startDocId, endDocId);
-    }
-
-    StarTreeDataTable dataSorter =
-        new StarTreeDataTable(file, dimensionSizeBytes, metricSizeBytes, getSortOrder());
-    dataSorter.sort(startDocId, endDocId, 0, dimensionSizeBytes);
-    if (debugMode) {
-      LOG.info("AFTER SORTING");
-      printFile(file, startDocId, endDocId);
+  private void sort(final int startDocId, int endDocId)
+      throws Exception {
+    List<Integer> recordIdList = new ArrayList<>();
+    for (int i = startDocId; i < endDocId; i++) {
+      recordIdList.add(i - startDocId);
     }
-  }
 
-  private int[] getSortOrder() {
-    if (sortOrder == null) {
-      sortOrder = new int[dimensionNames.size()];
-      for (int i = 0; i < dimensionsSplitOrder.size(); i++) {
-        sortOrder[i] = dimensionNameToIndexMap.get(dimensionsSplitOrder.get(i));
-      }
-      // add remaining dimensions that were not part of dimensionsSplitOrder
-      int counter = 0;
-      for (String dimName : dimensionNames) {
-        if (!dimensionsSplitOrder.contains(dimName)) {
-          sortOrder[dimensionsSplitOrder.size() + counter] = dimensionNameToIndexMap.get(dimName);
-          counter = counter + 1;
+    // Sort the record id list according to sort order.
+    Collections.sort(recordIdList, new Comparator<Integer>() {
+      @Override
+      public int compare(Integer docId1, Integer docId2) {
+        int pos1 = (docId1 + startDocId) * _totalSizeBytes;
```

Comment:
How will this work in presence of multi-valued columns, where the records may be of different sizes? If that is not supported, then may be add a Precondition somewhere appropriate.